pkg/hub/managedcluster/controller.go

```diff
@@ -78,26 +79,26 @@ func (c *managedClusterController) sync(ctx context.Context, syncCtx factory.Syn
 
 	managedCluster = managedCluster.DeepCopy()
 	if managedCluster.DeletionTimestamp.IsZero() {
-		hasFinalizer := false
-		for i := range managedCluster.Finalizers {
-			if managedCluster.Finalizers[i] == managedClusterFinalizer {
-				hasFinalizer = true
-				break
+		if !controllerutil.ContainsFinalizer(managedCluster, managedClusterFinalizer) {
+			controllerutil.AddFinalizer(managedCluster, managedClusterFinalizer)
+			if _, err := c.clusterClient.ClusterV1().ManagedClusters().Update(ctx, managedCluster, metav1.UpdateOptions{}); err != nil {
+				return err
 			}
 		}
-		if !hasFinalizer {
-			managedCluster.Finalizers = append(managedCluster.Finalizers, managedClusterFinalizer)
-			_, err := c.clusterClient.ClusterV1().ManagedClusters().Update(ctx, managedCluster, metav1.UpdateOptions{})
-			return err
-		}
 	}
 
 	// Spoke cluster is deleting, we remove its related resources
 	if !managedCluster.DeletionTimestamp.IsZero() {
 		if err := c.removeManagedClusterResources(ctx, managedClusterName); err != nil {
 			return err
 		}
-		return c.removeManagedClusterFinalizer(ctx, managedCluster)
+		if controllerutil.ContainsFinalizer(managedCluster, managedClusterFinalizer) {
+			controllerutil.RemoveFinalizer(managedCluster, managedClusterFinalizer)
```

Comment:
I hasn't take a look at the details of RemoveFinalizer, will it ensure the ordering after a deletion?

Reply:
Sorry, I don't understand the ordering meaing, did you say the order of finalizes? Why we need order? `RemoveFinalizer` will not change finalize order.

```
			f = append(f[:i], f[i+1:]...)

```

Reply:
yes, the ordering of the finalizers. Some controllers may rely on the ordering of the finalizers, it is not normal but it happens.

Reply:
RemoveFinalizer will keep the order. By the way, could you specify which controller use this feature?

Reply:
not in registration, but an external consumer (controller) could do that.